---

- name: Check rclone authentication
  ansible.builtin.uri:
    url: http://{{ rclone_host }}:{{ rclone_port }}/operations/stat
    body:
      fs: "{{ remote_name }}"
      remote: '/'
    body_format: json
    method: POST
    url_username: "{{ rclone_username }}"
    url_password: "{{ rclone_password }}"
    return_content: true
    headers:
      Accept: application/json
  register: rclone_auth_check
  changed_when: rclone_auth_check.status == 200
  failed_when:
    - rclone_auth_check.json.error is defined
    - "'empty token found' not in rclone_auth_check.json.error"

- name: Retry loop
  when: "'empty token found' in rclone_auth_check.json.error | default('')"
  block:
    - name: Authenticate Rclone Notice
      ansible.builtin.debug:
        msg: |
          ======== NOTICE ========
          Rclone needs to be authenticated. Please follow these steps:
          * Connect to {{ rclone_host }}
          * Run: sudo rclone config reconnect {{ remote_name }}:
          * DO NOT use auto config.
          * Follow instructions in output.
          * Resume when completed.

    - name: Wait for authentication
      ansible.builtin.pause:

    - name: Re-check rclone authentication
      ansible.builtin.uri:
        url: http://{{ rclone_host }}:{{ rclone_port }}/operations/about
        body:
          fs: "{{ remote_name }}"
        body_format: json
        method: POST
        url_username: "{{ rclone_username }}"
        url_password: "{{ rclone_password }}"
        return_content: true
        headers:
          Accept: application/json
      register: rclone_auth_recheck
      changed_when: rclone_auth_recheck.status == 200
      failed_when: "'empty token found' in rclone_auth_recheck.json.error | default('')"

  rescue:
    - name: Authentication failed
      ansible.builtin.debug:
        msg: Authentication failed, please retry

    - name: Include task
      ansible.builtin.include_tasks:
        file: rclone-authenticate-remote.yml
